core/node/node_framework/src/service/tests.rs

```diff
@@ -0,0 +1,232 @@
+        "wire_error_layer"
+    }
+
+    async fn wire(self: Box<Self>, _node: ServiceContext<'_>) -> Result<(), WiringError> {
+        Err(WiringError::Internal(anyhow!("wiring error")))
+    }
+}
+
+// ZkStack Service's `run()` method has to take into account errors on wiring step.
+#[test]
+fn test_run_with_error_tasks() {
+    let mut zk_stack_service = ZkStackService::new().unwrap();
+    let error_layer = WireErrorLayer;
+    zk_stack_service.add_layer(error_layer);
+    let result = zk_stack_service.run();
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        "One or more task weren't able to start".to_string(),
+        "Incorrect result for creating a service with wire error layer"
+    );
+}
+
+// ZkStack Service's `run()` method has to take into account errors on wiring step.
+#[derive(Debug)]
+struct TaskErrorLayer;
+
+#[async_trait::async_trait]
+impl WiringLayer for TaskErrorLayer {
+    fn layer_name(&self) -> &'static str {
+        "task_error_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(ErrorTask));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ErrorTask;
+
+#[async_trait::async_trait]
+impl Task for ErrorTask {
+    fn name(&self) -> &'static str {
+        "error_task"
+    }
+    async fn run(self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        anyhow::bail!("error task")
+    }
+}
+
+// ZkStack Service's `run()` method has to take into account errors inside task execution.
+#[test]
+fn test_run_with_failed_tasks() {
+    let mut zk_stack_service: ZkStackService = ZkStackService::new().unwrap();
+    zk_stack_service.add_layer(TaskErrorLayer);
+    let result = zk_stack_service.run();
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        "Task error_task failed".to_string(),
+        "Incorrect result for creating a service with task that fails"
+    );
+}
+
+// Layer that adds a task that updates resource and returns
+#[derive(Debug)]
+struct SuccessfulTaskLayer(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl WiringLayer for SuccessfulTaskLayer {
+    fn layer_name(&self) -> &'static str {
+        "successful_task_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(SuccessfulTask(self.0.clone())));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct SuccessfulTask(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl Task for SuccessfulTask {
+    fn name(&self) -> &'static str {
+        "successful_task"
+    }
+    async fn run(self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        let mut guard = self.0.lock().unwrap();
+        *guard = true;
+        Ok(())
+    }
+}
+
+// Layer for task that updates resource and returns only after stop signal received
+#[derive(Debug)]
+struct RemainingTaskLayer(Arc<Mutex<bool>>);
```

Comment:
Not sure I understand

Reply:
You have `SuccessfulTaskLayer` and `RemainingTaskLayer`, but this seems to be excessive. You can add two tasks from a single layer.

Reply:
If I add 2 tasks, `RemainingTask` and `SuccessfulTask`, into one layer, I cannot check that the flag has changed. 
I mean...it will not be possible to check a) the flag changed successfully  to `true` after `SuccessfulTask` finish.
Because  `RemainingTask` will be run after stop signal and again flag value will be false.

Reply:
You can have two fields in your layer (e.g. `Arc<Mutex<bool>>` for the first task and `Arc<Mutex<bool>>` for the second task).

Reply:
you are right..